In [2]:
import pandas as pd
import numpy as np
from sklearn.linear_model import LinearRegression

In [3]:
df = pd.read_csv("Ecom Expense.csv")

In [4]:
df.head()

Transaction ID  Age    Items   Monthly Income  Transaction Time  Record  \
0         TXN001    42       10            7313        627.668127       5   
1         TXN002    24        8           17747        126.904567       3   
2         TXN003    47       11           22845        873.469701       2   
3         TXN004    50       11           18552        380.219428       7   
4         TXN005    60        2           14439        403.374223       2   

   Gender City Tier  Total Spend  
0  Female    Tier 1  4198.385084  
1  Female    Tier 2  4134.976648  
2    Male    Tier 2  5166.614455  
3  Female    Tier 1  7784.447676  
4  Female    Tier 2  3254.160485

In [5]:
dummy_gender = pd.get_dummies(df["Gender"], prefix="Gender") ## creacion de las variables dummy de la columna "Gender" del df
dummy_city_tier = pd.get_dummies(df["City Tier"], prefix="City") ## creacion de las variables dummy de la columna "City Tier" del df

In [6]:
dummy_gender.head()

Gender_Female  Gender_Male
0              1            0
1              1            0
2              0            1
3              1            0
4              1            0

In [7]:
dummy_city_tier.head()

City_Tier 1  City_Tier 2  City_Tier 3
0            1            0            0
1            0            1            0
2            0            1            0
3            1            0            0
4            0            1            0

In [8]:
columns_names = df.columns.values.tolist() ##obtengo el nombre de las columnas del dataset
columns_names

['Transaction ID',
 'Age ',
 ' Items ',
 'Monthly Income',
 'Transaction Time',
 'Record',
 'Gender',
 'City Tier',
 'Total Spend']

In [9]:
df_new = df[columns_names].join(dummy_gender) ##hago una union de las columnas del dataframe que ya tengo con las columnas del dummy dataframe Gender
df_new.head()

Transaction ID  Age    Items   Monthly Income  Transaction Time  Record  \
0         TXN001    42       10            7313        627.668127       5   
1         TXN002    24        8           17747        126.904567       3   
2         TXN003    47       11           22845        873.469701       2   
3         TXN004    50       11           18552        380.219428       7   
4         TXN005    60        2           14439        403.374223       2   

   Gender City Tier  Total Spend  Gender_Female  Gender_Male  
0  Female    Tier 1  4198.385084              1            0  
1  Female    Tier 2  4134.976648              1            0  
2    Male    Tier 2  5166.614455              0            1  
3  Female    Tier 1  7784.447676              1            0  
4  Female    Tier 2  3254.160485              1            0

In [10]:
columns_names = df_new.columns.values.tolist() ## vuelvo a obtener las columnas, esta vez del dataframe nuevo que es el que ya tiene asignadas las columnas de Gender.

In [11]:
df_new = df_new[columns_names].join(dummy_city_tier) ##hago una union de las columnas del dataframe que ya tengo con las columnas del dummy dataframe citi_tier


In [12]:
df_new.head() ## df final

Transaction ID  Age    Items   Monthly Income  Transaction Time  Record  \
0         TXN001    42       10            7313        627.668127       5   
1         TXN002    24        8           17747        126.904567       3   
2         TXN003    47       11           22845        873.469701       2   
3         TXN004    50       11           18552        380.219428       7   
4         TXN005    60        2           14439        403.374223       2   

   Gender City Tier  Total Spend  Gender_Female  Gender_Male  City_Tier 1  \
0  Female    Tier 1  4198.385084              1            0            1   
1  Female    Tier 2  4134.976648              1            0            0   
2    Male    Tier 2  5166.614455              0            1            0   
3  Female    Tier 1  7784.447676              1            0            1   
4  Female    Tier 2  3254.160485              1            0            0   

   City_Tier 2  City_Tier 3  
0            0            0  
1            1            0  
2            1            0  
3            0            0  
4            1            0

# la propuesta consistirá en predecir el Total Spend (gasto total) 

In [13]:
columnas_predictoras = ["Monthly Income","Transaction Time","Record","Gender_Female","Gender_Male","City_Tier 1","City_Tier 2","City_Tier 3"]

In [14]:
X = df_new[columnas_predictoras]
Y = df_new["Total Spend"]

In [21]:
##ejecucion del modelo
lm = LinearRegression() 
lm.fit(X,Y)
df_new["prediccionTotalSpend"] = lm.predict(pd.DataFrame(df_new[columnas_predictoras]))
df_new.head()

Transaction ID  Age    Items   Monthly Income  Transaction Time  Record  \
0         TXN001    42       10            7313        627.668127       5   
1         TXN002    24        8           17747        126.904567       3   
2         TXN003    47       11           22845        873.469701       2   
3         TXN004    50       11           18552        380.219428       7   
4         TXN005    60        2           14439        403.374223       2   

   Gender City Tier  Total Spend  Gender_Female  Gender_Male  City_Tier 1  \
0  Female    Tier 1  4198.385084              1            0            1   
1  Female    Tier 2  4134.976648              1            0            0   
2    Male    Tier 2  5166.614455              0            1            0   
3  Female    Tier 1  7784.447676              1            0            1   
4  Female    Tier 2  3254.160485              1            0            0   

   City_Tier 2  City_Tier 3  prediccionTotalSpend  
0            0            0           4903.696720  
1            1            0           4799.434826  
2            1            0           5157.082504  
3            0            0           8068.012996  
4            1            0           3581.980335

In [22]:
lm.intercept_ ##obtiene el valor del intercepto(es decir el valor de alfa)

-79.41713030137089

In [23]:
list(zip(columnas_predictoras, lm.coef_)) ##obtiene el valor del los coeficientes de cada columnas predictora (es decir el valor de las betas)

[('Monthly Income', 0.14753898049205735),
 ('Transaction Time', 0.15494612549589615),
 ('Record', 772.2334457445644),
 ('Gender_Female', -131.02501325554636),
 ('Gender_Male', 131.02501325554644),
 ('City_Tier 1', 76.76432601049524),
 ('City_Tier 2', 55.13897430923252),
 ('City_Tier 3', -131.90330031972772)]

In [24]:
lm.score(X,Y) ## valor de R2

0.9179923586131016

#### el modelo puede ser escrito como:

df_new["prediccionTotalSpend"] = -79.41713030137089 + (df_new["Monthly Income"] * 0.14753898049205735) + (df_new["Transaction Time"] * 0.15494612549589615) + (df_new["Record"] * 772.2334457445644) + (df_new["Gender_Female"] * (-131.02501325554636)) + (df_new["Gender_Male"] * 131.02501325554644) + (df_new["City_Tier 1"] * 76.76432601049524) + (df_new["City_Tier 2"] * 55.13897430923252) + (df_new["City_Tier 3"] * (-131.90330031972772))

In [20]:
##calculo SSD
SSD = np.sum((df_new["prediccionTotalSpend"] - df_new["Total Spend"])**2)
SSD

##calculo del RSE
n = len(df_new)
k = len(columnas_predictoras)
RSE = np.sqrt(SSD/n-k-1)
RSE

##calculo del promedio de total spend
TotalSpend_mean = np.mean(df_new["Total Spend"])
TotalSpend_mean

##calculo del error
error = (RSE/TotalSpend_mean)
error

0.13006194434791704